### ISSUE TITLE
Huge mismatch in the calculation of the offset when using estimation type 2 and 3

### ISSUE THREAD

**Issue by fjandrad** [robotology-playground/insitu-ft-analysis#41]
When we use estimation type 2
We substract the mean to the information we use to calibrate and then we calculate the offset
```
/hat{r}=r-/mu r
/hat{w}=w-/mu w
```
and offset is then
`
o=C^*/mu r - /mu w`
In the paper in model based insitu we claimed that doing this should still result in a good estimation of the offset.

When using estimation type 3 we perform the one shot technique : 
We augment the raw data with 1 so that it will automatically will calculate the offset along with the calibration matrix.
In the case of only regularization and offset the give pretty much the same amount of offset. 
But when adding the temperature information then the offset calculation differs alot

In type 3 seems to have worst offset estimation in a grid data which has reasonable excitation.
The question is why? and how to deal with this so that we can trust the oneshot to give us a reasonable offset with temperature

**UPDATE** : bug was fund in use linear model to calibrate script. Index of temperature was wrong and seemed to be changing the lambda value.

**Comment by fjandrad** [comment]
A possible reason for this could be that we are introducing more unknowns as we are trying to fit also the temperature giving rise to multiple solutions that can variate between the offset and the temperature coefficients since they do not have a  specific reference. This could also explain why type 2 works better than type 3 for the grid case `2018_07_10_Grid`. When testing with a dataset as tz that alone does not excite enough the sensors both type 2 and 3 give clearly wrong estimates for the offset even if the result is completly the opposite ( huge offset for type 2 and almost no offset for type 3 ) even if the resulting predicted measures have low error.

A possible assumption is just to take the first value or first couple of values and appli ftdata-estiamteddata to obtain a sort of  reference offset.

**Comment by fjandrad** [comment]
## Consequences of proposed solution
With the proposed solution ( add a previous information on the offset for the one shot estimation).
Given that using this previous information without temperature seems to give a worst MSE as shown below we chose to apply only to the cases temperature coefficients are also estimated:
MSE applying assumption on all one shot cases:

| With Temperature | Fx       | Fy      | Fz     | Tx     | Ty     | Tz     | Type of estimation |
|------------------|----------|---------|--------|--------|--------|--------|--------------------|
| No               | 3.4952   | 6.2629  | 8.2695 | 0.0655 | 0.0744 | 0.0202 | 1                  |
| No               | 4.8280   | 6.0875  | 7.8342 | 0.0645 | 0.0744 | 0.0215 | 2                  |
| No               | 5.3752   | 5.9217  | 7.4268 | 0.0653 | 0.0741 | 0.0210 | 3                  |
| No               | 198.9299 | 12.4058 | 8.8322 | 0.0667 | 0.0816 | 0.0345 | 4                  |
| Yes              | 3.3360   | 5.8737  | 3.7769 | 0.0639 | 0.0740 | 0.0199 | 1                  |
| Yes              | 3.2815   | 7.6185  | 3.7675 | 0.0641 | 0.0741 | 0.0191 | 2                  |
| Yes              | 3.3115   | 6.0288  | 3.7750 | 0.0639 | 0.0740 | 0.0200 | 3                  |
| Yes              | 3.7709   | 6.0632  | 4.7204 | 0.0639 | 0.0887 | 0.0239 | 4                  |

MSE with previous offset only when using temperature:

| With Temperature | Fx     | Fy     | Fz     | Tx     | Ty     | Tz     | Type of estimation |
|------------------|--------|--------|--------|--------|--------|--------|--------------------|
| No               | 3.4952 | 6.2629 | 8.2695 | 0.0655 | 0.0744 | 0.0202 | 1                  |
| No               | 4.8280 | 6.0875 | 7.8342 | 0.0645 | 0.0744 | 0.0215 | 2                  |
| No               | 4.8280 | 6.0875 | 7.8341 | 0.0645 | 0.0744 | 0.0215 | 3                  |
| No               | 3.3580 | 6.0099 | 3.8136 | 0.0639 | 0.0740 | 0.0198 | 4                  |
| Yes              | 3.3360 | 5.8737 | 3.7769 | 0.0639 | 0.0740 | 0.0199 | 1                  |
| Yes              | 3.2815 | 7.6185 | 3.7675 | 0.0641 | 0.0741 | 0.0191 | 2                  |
| Yes              | 3.3115 | 6.0288 | 3.7750 | 0.0639 | 0.0740 | 0.0200 | 3                  |
| Yes              | 3.7709 | 6.0632 | 4.7204 | 0.0639 | 0.0887 | 0.0239 | 4                  | 

## Results in the offset estimation:
we get the following results:
 **initial offset in measurements**:
` -84.8463   17.7822   10.9352    0.3921    1.2137   -0.7441`

**offset in using workbenchMatrix*rawEstimatedOffset**

| With Temperature | Fx        | Fy      | Fz       | Tx      | Ty     | Tz      | Type of estimation |
|------------------|-----------|---------|----------|---------|--------|---------|--------------------|
| No               | -87.0664  | 16.2965 | 2.6376   | 0.4095  | 1.0429 | -0.7039 | 1                  |
| No               | -83.7846  | 21.3598 | 3.2740   | 0.4136  | 1.0635 | -0.7943 | 2                  |
| No               | -83.7846  | 21.3598 | 3.2740   | 0.4136  | 1.0635 | -0.7943 | 3                  |
| No               | -88.4086  | 19.7256 | 7.8302   | 0.3489  | 1.0296 | -0.6982 | 4                  |
| Yes              | -87.0664  | 16.2965 | 2.6376   | 0.4095  | 1.0429 | -0.7039 | 1                  |
| Yes              | -73.7225  | 61.9936 | 7.3069   | 0.8094  | 1.3323 | -0.7444 | 2                  |
| Yes              | -83.7846  | 21.3598 | 3.2740   | 0.4136  | 1.0635 | -0.7943 | 3                  |
| Yes              | -120.3109 | 9.5899  | -62.4545 | -1.1475 | 7.7757 | -1.6068 | 4                  |


**Remaining offset by subtracting the estimated offset in the measured ft data**


| With Temperature | Fx       | Fy      | Fz       | Tx      | Ty      | Tz      | Type of estimation |
|------------------|----------|---------|----------|---------|---------|---------|--------------------|
| No               | -2.2201  | -1.4857 | -8.2977  | 0.0174  | -0.1708 | 0.0402  | 1                  |
| No               | 1.0617   | 3.5776  | -7.6612  | 0.0215  | -0.1502 | -0.0502 | 2                  |
| No               | 1.0617   | 3.5776  | -7.6612  | 0.0215  | -0.1502 | -0.0502 | 3                  |
| No               | -3.5623  | 1.9434  | -3.1050  | -0.0432 | -0.1841 | 0.0460  | 4                  |
| Yes              | -2.2201  | -1.4857 | -8.2977  | 0.0174  | -0.1708 | 0.0402  | 1                  |
| Yes              | 11.1238  | 44.2113 | -3.6283  | 0.4173  | 0.1186  | -0.0003 | 2                  |
| Yes              | 1.0617   | 3.5776  | -7.6612  | 0.0215  | -0.1502 | -0.0502 | 3                  |
| Yes              | -35.4646 | -8.1923 | -73.3897 | -1.5396 | 6.5620  | -0.8626 | 4                  |

**Comment by traversaro** [comment]
I tried to understand the problem from the issue but it is not trivial, when you are back in IIT I think we  can discuss about it.

**Comment by fjandrad** [comment]
From taking a look at the results again and the formulations. It seems that estimations 1 and 2 impose the assumption that the offset is one and everything else can be taken care of by the calibration  + temperature.
This comes from the fact that the offset is somewhat removed before attempting the calibration.
In these cases the coefficients of the temperature are small and work as a minor correction of the offset later in the experiment where the temperature has increased. Which is congruent with our original assumption that the temperature mainly affected the temperature. 
For type 3 and 4 the raw data is not manipulated at all before attempting the calibration. In these estimations the coefficient of the temperature is way higher than the other types. Which would suggest that the temperature is more critical than initially thought. 
But taking into account the MSE performance and the results from the tests in [the temperature tests](https://github.com/loc2/component_ft-sensors/issues/28#issuecomment-415392889), this suggests that type 3 and 4 are mainly over fitting. In these types also the offset is hard to understand and compare to how we consider the offsets in the other cases.

When running the estimation of calibration with lambda 1 we get:
| Temperature coefficients | type 1  | type 2  | type 3  | type 4  |
|--------------------------|---------|---------|---------|---------|
| Fx                       | 0.0481  | 0.1106  | 0.2710  | -0.5646 |
| Fy                       | -0.0515 | 0.0071  | 0.2043  | 0.0198  |
| Fz                       | -0.1460 | -0.1604 | 1.2714  | -1.7722 |
| Tx                       | 0.0020  | 0.0015  | 0.0327  | -0.0370 |
| Ty                       | 0.0024  | 0.0019  | -0.0340 | 0.1894  |
| Tz                       | -0.0008 | -0.0021 | 0.0368  | -0.0345 |

For the time being I would consider to change type 3 to instead of one shot to be something like 2 but all data put together and keep 4 as is. Even if for now they seem to be over fitting and the offset different from expected, I will perform the other tests for comparing performance before fully discarding.

In conclusion in type 1 and 2 we enforce a bit our assumption that temperature acts more on the offset and that we can estimate the offset independently. Type 4 makes no such assumption and allows extra freedom for over fitting which might indicate 2 things either temperature is more critical than originally thought or using that type of estimation requires more different values of temperature to avoid over fitting as much in that local solution.

@traversaro this is what I came up after double checking all of it.

**Comment by traversaro** [comment]
>  the temperature mainly affected the temperature.

I guess you meant "the temperature mainly affected the **offset**"

**Comment by fjandrad** [comment]
I will close this issue, it will become more clear when we study the offsets themselves. It seems that is mainly the one shot having some weird offset. It might be because is trying to group the offset in the temperature with the one on the calibration matrix.
Given that the temperature is slow I think is valid to just substract the first value and then keep as reference later on. I will test this and show results in another issue.
